ADR/partner-tasks.md

```diff
@@ -0,0 +1,34 @@
+# Partner Tasks in Build/Test Pipelines
+
+Date: 2023-03-06
+
+## Status
+
+Provisional
+
+## Context
+
+* As a Red Hat Partner, I would like to offer our service's capability as a Tekton Task that would be executed in a user's build/test Pipeline on StoneSoup.
+* As a StoneSoup maintainer, I would like to provide a way to ensure the above Tekton Task is validated and approved for use in the build/test Pipeline on StoneSoup.
+
+Associated business ask
+https://issues.redhat.com/browse/STONE-549
+
+## Decision
+
+### Plumbing
+
+1. Setup a new Git repository.
+2. Define a directory structure for Task submissions as manifests in the yaml format.
+3. Configure a CI job that validates the Tasks upon opening of a pull request.
+4. Configure a CI job that generates an OCI artifact consumable in a Tekton Pipeline.
+
+### Day-to-day operations
+1. Partner opens a PR with a new/updated Task.
+2. Upon approval from the Build/Test team, the changes are merged.
+3. Manually/Automatically, a PR is opened to the https://github.com/redhat-appstudio/build-definitions repository
+
+## Consequences
+
+* We have a mechanism to take in Tekton Tasks from Partners with due diligence.
```

Comment:
What would certification of these tasks look like? For inclusion in Red Hat's pipleine? For inclusion in others' pipelines? Can we provide any clarity around "due diligence"?

The proposed SLSA v1.0 is largely just a transference of trust to the build system. A "similar" build system basically has trusted build tasks
> The way we defined the core elements of a "secure" build service in the Secure Software Factory Ref Arch is pretty much as follows:
> * Secure orchestration of the build pipeline. i.e. only approved pipelines that follow established policy and rules should be allowed to run.
> * Secure orchestration of the build workload. i.e. only approved build workloads are allowed to be run, and if someone tampers with the orchestration of a workload it should be detected or prevented. An example here is using SPIFFE or other workload identities.
> * Secure the actions in the build workload itself. i.e. a build that tries to perform malicious actions should be prevented and the build workload should be isolated enough to limit blast radius.

Reply:
The exhaustive list of "due diligence" is being built out by the @vdemeester 's team ( tekton ecosystem / certification ). 

I doubt we'd need to have that in our initial iteration, therefore our stripped down list, which we should iterate on:

* No privilege escalation / or no requirements that need unreasonable privileges.
* Should be integratabtle into the build pipeline, ie, works with the inputs/outputs we have today.
* Should work with reasonable defaults.
* Should be skip-able if credentials/tokens are missing.

Reply:
I would expect some level of "due diligence" to be provided by Red hat as well -- in addition to tekton ecosystem certification. I know that HACBS always has an 80/20 rule where the out-of-the-box configuration was only intended to solve 80% of use cases. The flexibility of the platform allowed for the other 20% to get through.

Therefore, partner tasks might not just be external partners, but also potentially internal users. How can we (via `ec`) certify that specific tasks are "approved" for use in software that we use? How can external users certify that specific tasks meet their requirements as well?

Even if we just have a stripped down list of diligence, I think we should include _something_ on it in the ADR, also mentioning how `ec` might be able to filter. @simonbaird, fyi on the ec angle.

> Should be skip-able if credentials/tokens are missing.

Would the tasks be injected into the pipeline by default? If so, this will produce many extraneous "skipped" tasks in the UI, especially as the number of partner tasks increases.

Reply:
Yes, I do expect internal tasks to go here as well. 

Yes, the due diligence is by Red Hat - in the form of a CI job that validates contributions made to the repository.

Reply:
Let's type out exactly what we mean versus the ambiguous "due diligence".  For instance:
Partner is able to test their Task in their own workspace by using a custom pipeline and modifying the contents of their .tekton/ directory.
The Partner can maintain a blog that identifies how their Task should be integrated into a user's pipelinerun definition. Instructions should include how to add a secret/token to activate the task.

Questions:
Is it possible for the Partner to use  Stonesoup as their CI solution for their Task?

Reply:
Even if Red Hat is "okay" with these partner-supplied tasks, all customers/users might not be okay with them for their own policies without additional diligence done. I think that we need to make sure we are transparent about what diligence happens (initially and long-term as updates are pushed) and what policies might be able to be applied to filter out these tasks.

Reply:
> Yes, the due diligence is by Red Hat - in the form of a CI job that validates contributions made to the repository.

Ideally, the Task that would get accepted and "certified" in the TKNECO realm, would pass a lot of gates ; best practices, linting, scaning, … and be heavily tested.. etc.. All this is has yet to be properly defined but that's the idea.

Reply:
* See the section "Definition of a valid Task"
* @burrsutter , see the "out-of-scope" section. 

let me know!